File: ruby/lib/phonetics/confusion.rb
# frozen_string_literal: true

require 'set'
require_relative '../phonetics'

# Listener-confusion distance over IPA phonemes.
#
# Where Phonetics::Levenshtein gives the strict edit-distance between two
# phonemic sequences using a weighted per-phoneme cost, this module gives the
# *perceptual* distance: how likely a native English listener is to confuse
# the two streams. Calibrated against Mad Gab puzzle data, which encodes a
# real natural-speech phenomenon: humans systematically resyllabify a stream
# of phones into a different lexical sequence and need only a soft set of
# acoustic cues to bridge between the two.
#
# Three modifications to standard weighted Levenshtein:
#
#   1. AFFINE GAP PENALTIES (Gotoh 1982). Opening a new alignment gap costs
#      GAP_OPEN; extending an existing gap costs GAP_EXTEND, where
#      GAP_EXTEND << GAP_OPEN. Mad Gab clues typically preserve the target's
#      phoneme sequence and add filler phonemes at word boundaries — that
#      filler manifests as a single contiguous gap, which a fixed indel
#      penalty would over-charge but an affine penalty handles correctly.
#
#   2. WEAK-PHONEME INDEL DISCOUNT. /ə/, /h/, /ʔ/, /ɦ/ are routinely
#      inserted, dropped, or hallucinated in casual English speech. Their
#      indel cost is WEAK_INDEL_COST, well below GAP_OPEN. This is the
#      mechanism by which "Hits Justice Dupe Hid Came" can absorb a non-
#      existent /h/ at every word start without much penalty.
#
#   3. EXPLICIT TWO-TIER API. Callers who want the raw acoustic distance
#      (clustering accents, comparing dialect transcriptions, modelling
#      ASR error) use Phonetics::Levenshtein. Callers who want the
#      perceptual judgement (Mad Gab solving, pun detection, mondegreen
#      analysis, mishearings) use Phonetics::Confusion.
#
# Public surface:
#
#   Phonetics::Confusion.distance(ipa1, ipa2)
#   Phonetics::Confusion.similarity(ipa1, ipa2)  # 1.0 - normalised distance
#   Phonetics::RubyConfusion.new(...).distance   # for verbose tracing
module Phonetics
  module Confusion
    extend self

    # Cost of starting a new gap (insertion or deletion of one phoneme that
    # isn't preceded by another insertion/deletion in the same direction).
    # Slightly higher than the largest typical sub cost on a "close" pair
    # so the algorithm prefers matching over skipping when phonemes are
    # mildly similar.
    GAP_OPEN = 0.60

    # Cost of extending an already-open gap by one more phoneme. Below
    # GAP_OPEN (so multi-phoneme re-syllabification at word boundaries is
    # cheap), but high enough that an unrelated string can't trivially
    # "skip-gap" itself into alignment with the target. Calibrated against
    # the Mad Gab corpus: this is roughly the per-phoneme penalty needed
    # to keep an unrelated phrase from out-scoring a real clue.
    GAP_EXTEND = 0.25

    # Phonemes that are perceptually transparent at word boundaries in
    # English. Indels of these always use WEAK_INDEL_COST regardless of
    # whether they're opening or extending a gap. The set is informed by
    # English speech-reduction literature (Bybee, Fokes & Bond) — these
    # are the segments with the highest measured deletion/insertion rates
    # in natural conversational speech.
    WEAK_PHONEMES = Set.new(%w[ə h ʔ ɦ]).freeze
    WEAK_INDEL_COST = 0.15

    # The word-boundary token is even cheaper to indel than a weak phoneme.
    # Mad Gab is fundamentally a re-syllabification operation: the surface
    # form rearranges word boundaries without changing the underlying
    # phoneme stream. Charging meaningfully for a boundary indel would
    # punish exactly the operation we want to track.
    BOUNDARY_INDEL_COST = 0.02

    # Empirically-confusable phoneme pairs whose acoustic distance under
    # Phonetics.distance overstates the perceptual gap. Drawn from the
    # Miller-Nicely (1955) consonant-confusion lineage and English-
    # variation studies (th-stopping, intervocalic /t/-flapping, the
    # canonical l/r confusion).
    #
    # The overlay is one-directional in the look-up but symmetric in
    # interpretation: presence of the pair {a,b} sets sub_cost(a,b) and
    # sub_cost(b,a) both to the overlay value.
    #
    # Strict Phonetics.distance is unaffected — only the Confusion path
    # consults this table.
    EMPIRICAL_OVERLAY = {
      # Th-stopping: /θ/→/t/, /ð/→/d/ (very common in casual speech,
      # AAVE, NYC English, and L2 English).
      Set[ 'θ', 't' ]  => 0.18,
      Set[ 'ð', 'd' ]  => 0.18,
      # Th-stopping into sibilants: less common but documented.
      Set[ 'θ', 's' ]  => 0.12,
      Set[ 'ð', 'z' ]  => 0.12,
      # Fric-stop crossing at the same place (the "p/f", "b/v" cousins).
      Set[ 'p', 'f' ]  => 0.20,
      Set[ 'b', 'v' ]  => 0.20,
      Set[ 't', 's' ]  => 0.20,
      Set[ 'd', 'z' ]  => 0.20,
      # Intervocalic /t/-flapping in American English: 'butter' ≈ 'budder'
      # ≈ 'buɾer'. The flap [ɾ] is acoustically nearly indistinguishable
      # from a brief voiced [d] between vowels.
      Set[ 't', 'ɾ' ]  => 0.10,
      Set[ 'd', 'ɾ' ]  => 0.05,
      # L/R confusion (universally high-confusion pair for L2 English;
      # the lateral fix already drops it to 0.10 acoustically, the
      # overlay pushes it slightly higher to reflect perceptual reality).
      Set[ 'l', 'ɹ' ]  => 0.15,
    }.freeze

    # Convenience: top-level entry point.
    def distance(ipa1, ipa2, verbose: false)
      RubyConfusion.new(ipa1, ipa2, verbose: verbose).distance
    end

    # 0..1 normalised similarity score. Worst case is one substitution per
    # position in the longer string, so dividing by max(len) gives a
    # bounded-range judgement comparable across phrase lengths.
    def similarity(ipa1, ipa2)
      a = Phonetics::String.new(ipa1).each_phoneme(boundaries: true).to_a
      b = Phonetics::String.new(ipa2).each_phoneme(boundaries: true).to_a
      max_n = [a.size, b.size].max
      return 1.0 if max_n.zero?

      d = RubyConfusion.new(ipa1, ipa2).distance
      [1.0 - (d.to_f / max_n), 0.0].max
    end

    # Per-phoneme substitution cost used by the confusion DP. Consults the
    # empirical overlay first, then falls back to Phonetics.distance, which
    # already handles compound phonemes (diphthongs/affricates), cross-class
    # bridges, and diacritics.
    def sub_cost(a, b)
      return 0.0 if a == b

      overlay = EMPIRICAL_OVERLAY[Set[a, b]]
      return overlay if overlay

      Phonetics.distance(a, b)
    end

    def weak?(phoneme)
      WEAK_PHONEMES.include?(phoneme) || phoneme == Phonetics::BOUNDARY_TOKEN
    end

    # Per-phoneme indel cost. The word-boundary token is at the bottom of
    # the cost ladder (re-syllabification is free); other weak phonemes
    # use WEAK_INDEL_COST; everything else falls back to the affine
    # GAP_OPEN/GAP_EXTEND machinery.
    def weak_indel_cost(phoneme)
      return BOUNDARY_INDEL_COST if phoneme == Phonetics::BOUNDARY_TOKEN
      return WEAK_INDEL_COST if WEAK_PHONEMES.include?(phoneme)

      nil
    end
  end

  # Pure-Ruby implementation of Confusion.distance. Mirrors the structure of
  # RubyLevenshtein so the two are interchangeable to callers that want a
  # verbose / introspectable distance computation.
  class RubyConfusion
    INF = Float::INFINITY

    attr_reader :a, :b

    def self.distance(a, b, verbose: false)
      new(a, b, verbose: verbose).distance
    end

    def initialize(a, b, verbose: false)
      @a = Phonetics::String.new(a).each_phoneme(boundaries: true).to_a
      @b = Phonetics::String.new(b).each_phoneme(boundaries: true).to_a
      @verbose = verbose
    end

    def distance
      m = a.size
      n = b.size
      return 0.0 if m.zero? && n.zero?
      return seed_cost(b) if m.zero?
      return seed_cost(a) if n.zero?

      # Three DP matrices for Gotoh's algorithm:
      #   mm[i][j] -- best score ending in a substitution/match at (i, j)
      #   xx[i][j] -- best score ending in a gap that consumes from a
      #   yy[i][j] -- best score ending in a gap that consumes from b
      mm = Array.new(m + 1) { Array.new(n + 1, INF) }
      xx = Array.new(m + 1) { Array.new(n + 1, INF) }
      yy = Array.new(m + 1) { Array.new(n + 1, INF) }
      mm[0][0] = 0.0

      # Seed the gap-only edges. The leftmost column models matching the
      # empty target against successive prefixes of `a`; the top row does
      # the symmetric thing for `b`. Both follow the open-then-extend rule,
      # with weak phonemes overriding both.
      (1..m).each do |i|
        ph = a[i - 1]
        prev = i == 1 ? 0.0 : xx[i - 1][0]
        xx[i][0] = prev + indel_step(ph, opening: i == 1)
      end
      (1..n).each do |j|
        ph = b[j - 1]
        prev = j == 1 ? 0.0 : yy[0][j - 1]
        yy[0][j] = prev + indel_step(ph, opening: j == 1)
      end

      (1..m).each do |i|
        ai = a[i - 1]
        a_weak = Confusion.weak?(ai)

        (1..n).each do |j|
          bj = b[j - 1]
          b_weak = Confusion.weak?(bj)

          # M: end in a match/mismatch. Can transition from any state.
          best_into_m = [mm[i - 1][j - 1], xx[i - 1][j - 1], yy[i - 1][j - 1]].min
          mm[i][j] = best_into_m + Confusion.sub_cost(ai, bj)

          # X: end in an a-consuming gap. Open a fresh gap from M or Y, or
          # extend an existing X-gap. Weak `a` overrides both costs.
          a_weak_cost = Confusion.weak_indel_cost(ai)
          xx[i][j] =
            if a_weak_cost
              [mm[i - 1][j], xx[i - 1][j], yy[i - 1][j]].min + a_weak_cost
            else
              [
                mm[i - 1][j] + Confusion::GAP_OPEN,
                xx[i - 1][j] + Confusion::GAP_EXTEND,
                yy[i - 1][j] + Confusion::GAP_OPEN,
              ].min
            end

          # Y: end in a b-consuming gap. Symmetric.
          b_weak_cost = Confusion.weak_indel_cost(bj)
          yy[i][j] =
            if b_weak_cost
              [mm[i][j - 1], xx[i][j - 1], yy[i][j - 1]].min + b_weak_cost
            else
              [
                mm[i][j - 1] + Confusion::GAP_OPEN,
                yy[i][j - 1] + Confusion::GAP_EXTEND,
                xx[i][j - 1] + Confusion::GAP_OPEN,
              ].min
            end
        end
      end

      print_matrix(mm, xx, yy) if @verbose
      [mm[m][n], xx[m][n], yy[m][n]].min
    end

    private

    # Edge case: comparing a non-empty string against the empty string. Cost
    # is one gap that extends across every phoneme of the non-empty side
    # (with weak phonemes still discounted individually).
    def seed_cost(phonemes)
      total = 0.0
      phonemes.each_with_index do |ph, i|
        total += indel_step(ph, opening: i.zero?)
      end
      total
    end

    def indel_step(phoneme, opening:)
      cost = Confusion.weak_indel_cost(phoneme)
      return cost if cost

      opening ? Confusion::GAP_OPEN : Confusion::GAP_EXTEND
    end

    def print_matrix(mm, xx, yy)
      puts "Confusion DP for a=#{a.inspect} b=#{b.inspect}"
      [['M', mm], ['X', xx], ['Y', yy]].each do |label, mat|
        puts "#{label} matrix:"
        mat.each { |row| puts '  ' + row.map { |v| v == INF ? '   ∞   ' : '%7.3f' % v }.join(' ') }
      end
    end
  end
end

File: rust/phonetics/src/confusion.rs
//! Listener-confusion distance — Gotoh affine-gap DP over IPA tokens,
//! with weak-phoneme indel discount, boundary-token discount, and an
//! empirical-confusion overlay on top of the acoustic substitution
//! cost.
//!
//! Where [`crate::levenshtein`] answers "how different are these two
//! waveforms?", this answers "will a listener parse one as the other?".
//! Calibrated against Mad Gab puzzle data and against the maintainer's
//! West Coast American English. The overlay is a hand-curated table;
//! the DP shape is from Gotoh (1982).
//!
//! Three matrices:
//!
//!   M[i][j]   best score ending in a match/substitution
//!   X[i][j]   best score ending in an a-consuming gap
//!   Y[i][j]   best score ending in a b-consuming gap
//!
//! Affine gap pricing: opening a gap costs `GAP_OPEN`, extending one
//! already open costs `GAP_EXTEND` << `GAP_OPEN`. Mad Gab clues
//! typically add filler phonemes at word boundaries; one long gap
//! costs `GAP_OPEN + (k-1) * GAP_EXTEND`, not `k * GAP_OPEN`.

use crate::{distance as phoneme_distance, symbols, tokenizer};

// ------------------------------------------------------------------
// Tuning constants
// ------------------------------------------------------------------

/// Cost of starting a new gap.
pub const GAP_OPEN: f64 = 0.60;

/// Cost of extending an already-open gap by one phoneme.
pub const GAP_EXTEND: f64 = 0.25;

/// Indel cost for "weak" phonemes — those routinely inserted, dropped,
/// or hallucinated in casual English: /ə/, /h/, /ʔ/, /ɦ/.
pub const WEAK_INDEL_COST: f64 = 0.15;

/// Indel cost for the word-boundary token. Lower than WEAK_INDEL_COST
/// because re-syllabification is the operation Mad Gab encodes; we
/// don't want to punish it.
pub const BOUNDARY_INDEL_COST: f64 = 0.02;

/// The four "weak" phonemes — segments with the highest measured
/// deletion/insertion rates in conversational English.
pub const WEAK_PHONEMES: &[&str] = &["ə", "h", "ʔ", "ɦ"];

/// True if `phoneme` is in the weak tier OR is the boundary token.
pub fn weak(phoneme: &str) -> bool {
    phoneme == symbols::BOUNDARY_TOKEN || WEAK_PHONEMES.contains(&phoneme)
}

/// Tier-appropriate indel cost for `phoneme`, or `None` if it should
/// fall back to the affine GAP_OPEN/GAP_EXTEND machinery.
pub fn weak_indel_cost(phoneme: &str) -> Option<f64> {
    if phoneme == symbols::BOUNDARY_TOKEN {
        Some(BOUNDARY_INDEL_COST)
    } else if WEAK_PHONEMES.contains(&phoneme) {
        Some(WEAK_INDEL_COST)
    } else {
        None
    }
}

// ------------------------------------------------------------------
// Empirical-confusion overlay
// ------------------------------------------------------------------
//
// Pairs whose acoustic distance under `crate::distance` overstates the
// perceptual gap. The table mixes cross-variety findings (Miller-Nicely
// 1955, generic English speech-perception studies) with American-
// English-specific mergers — most of the overlay is calibrated against
// a West Coast American baseline because that's the dialect the
// maintainer hears natively.

/// (sym_a, sym_b, perceptual_cost). The lookup treats pairs as
/// unordered: sub_cost(a, b) == sub_cost(b, a).
const OVERLAY: &[(&str, &str, f64)] = &[
    // Cross-variety
    ("θ", "t", 0.18),
    ("ð", "d", 0.18),
    ("θ", "s", 0.12),
    ("ð", "z", 0.12),
    ("p", "f", 0.20),
    ("b", "v", 0.20),
    ("t", "s", 0.20),
    ("d", "z", 0.20),
    ("l", "ɹ", 0.15),
    // American (esp. WCE)
    ("t", "ɾ", 0.10),
    ("d", "ɾ", 0.05),
    ("ɑ", "ɔ", 0.05),
    ("ɑ", "ɒ", 0.05),
    ("t", "ʔ", 0.08),
    ("d", "ʔ", 0.20),
    ("u", "y", 0.15),
    ("u", "ɯ", 0.15),
    ("u", "ʉ", 0.10),
    ("o", "ə", 0.20),
    ("ʌ", "ɑ", 0.10),
];

/// Returns the overlay cost for an unordered pair, if present.
fn overlay_cost(a: &str, b: &str) -> Option<f64> {
    for &(x, y, cost) in OVERLAY {
        if (x == a && y == b) || (x == b && y == a) {
            return Some(cost);
        }
    }
    None
}

/// Per-phoneme substitution cost used by the Confusion DP. Identity
/// short-circuits; then the overlay; then the acoustic metric.
pub fn sub_cost(a: &str, b: &str) -> f64 {
    if a == b {
        return 0.0;
    }
    if let Some(c) = overlay_cost(a, b) {
        return c;
    }
    phoneme_distance(a, b)
}

// ------------------------------------------------------------------
// Gotoh DP
// ------------------------------------------------------------------

/// Listener-confusion distance between two IPA strings.
pub fn distance(s1: &str, s2: &str) -> f64 {
    let a = tokenizer::tokens(s1, true);
    let b = tokenizer::tokens(s2, true);
    distance_from_tokens(&a, &b)
}

/// 0..1 normalised similarity score. Worst case is one substitution
/// per position in the longer string; dividing by max(len) gives a
/// bounded judgement comparable across phrase lengths.
pub fn similarity(s1: &str, s2: &str) -> f64 {
    let a = tokenizer::tokens(s1, true);
    let b = tokenizer::tokens(s2, true);
    let max_n = a.len().max(b.len());
    if max_n == 0 {
        return 1.0;
    }
    let d = distance_from_tokens(&a, &b);
    (1.0 - d / max_n as f64).max(0.0)
}

/// Confusion distance over pre-tokenised phoneme sequences.
#[allow(clippy::too_many_lines)]
pub fn distance_from_tokens<S: AsRef<str>>(a: &[S], b: &[S]) -> f64 {
    let m = a.len();
    let n = b.len();
    if m == 0 && n == 0 {
        return 0.0;
    }
    if m == 0 {
        return seed_cost(b);
    }
    if n == 0 {
        return seed_cost(a);
    }

    let width = n + 1;
    let cells = (m + 1) * width;
    let inf = INF;
    let mut mm = vec![inf; cells];
    let mut xx = vec![inf; cells];
    let mut yy = vec![inf; cells];
    mm[0] = 0.0;

    // Seed gap-only edges.
    for i in 1..=m {
        let ph = a[i - 1].as_ref();
        let step = indel_step(ph, i == 1);
        let prev = if i == 1 { 0.0 } else { xx[(i - 1) * width] };
        xx[i * width] = prev + step;
    }
    for j in 1..=n {
        let ph = b[j - 1].as_ref();
        let step = indel_step(ph, j == 1);
        let prev = if j == 1 { 0.0 } else { yy[j - 1] };
        yy[j] = prev + step;
    }

    for i in 1..=m {
        let ai = a[i - 1].as_ref();
        let a_weak_cost = weak_indel_cost(ai);
        for j in 1..=n {
            let bj = b[j - 1].as_ref();
            let b_weak_cost = weak_indel_cost(bj);

            let here     = i * width + j;
            let up       = (i - 1) * width + j;
            let left     = i * width + (j - 1);
            let diag     = (i - 1) * width + (j - 1);

            // M: end in match/mismatch.
            mm[here] = min3(mm[diag], xx[diag], yy[diag]) + sub_cost(ai, bj);

            // X: end in an a-consuming gap.
            xx[here] = if let Some(c) = a_weak_cost {
                min3(mm[up], xx[up], yy[up]) + c
            } else {
                min3(mm[up] + GAP_OPEN, xx[up] + GAP_EXTEND, yy[up] + GAP_OPEN)
            };

            // Y: end in a b-consuming gap.
            yy[here] = if let Some(c) = b_weak_cost {
                min3(mm[left], xx[left], yy[left]) + c
            } else {
                min3(mm[left] + GAP_OPEN, yy[left] + GAP_EXTEND, xx[left] + GAP_OPEN)
            };
        }
    }

    let last = m * width + n;
    min3(mm[last], xx[last], yy[last])
}

// ------------------------------------------------------------------
// Helpers
// ------------------------------------------------------------------

const INF: f64 = 1e18;

#[inline]
fn min3(a: f64, b: f64, c: f64) -> f64 {
    a.min(b).min(c)
}

fn indel_step(phoneme: &str, opening: bool) -> f64 {
    if let Some(c) = weak_indel_cost(phoneme) {
        return c;
    }
    if opening { GAP_OPEN } else { GAP_EXTEND }
}

fn seed_cost<S: AsRef<str>>(tokens: &[S]) -> f64 {
    let mut total = 0.0;
    for (i, ph) in tokens.iter().enumerate() {
        total += indel_step(ph.as_ref(), i == 0);
    }
    total
}

#[cfg(test)]
mod tests {
    use super::*;

    const EPS: f64 = 1e-12;

    #[test]
    fn matches_ruby_reference_distances() {
        // Pure-Ruby reference values (f64).
        let cases: &[(&str, &str, f64)] = &[
            ("kæt", "kæt", 0.0),
            ("kæt", "kʌt", 0.145_085_455_502_268_37),
            ("ɪtsdʒʌstəstupɪdgeɪm",
             "hɪtsdʒʌstɪsduphɪdkeɪm",
             0.769_519_814_165_789_6),
            ("ɪtsdʒʌstəstupɪdgeɪm",
             "jɔrmʌðɝwɛrzsneɪkɝz",
             6.485_176_104_558_604),
            ("æpəlpaɪ", "eɪppʊlpaɪ", 1.047_133_181_946_413_6),
            ("nidəkɔfi", "nidɑkhɔffi", 0.968_204_431_378_303_3),
            ("aɪlʌvju", "aɪlʌvju", 0.0),
            ("ɪts dʒʌst", "ɪt sdʒʌst", 0.04),
            ("stupɪd", "stupɪdli", 0.85),
            ("stupɪd", "hstupɪd", 0.15),
            ("", "", 0.0),
            ("kæt", "", 1.1),
        ];

        for (a, b, expected) in cases {
            let got = distance(a, b);
            assert!(
                (got - expected).abs() < EPS,
                "distance({a:?}, {b:?}) = {got}, expected {expected}",
            );
        }
    }

    #[test]
    fn empirical_overlay_fires_on_t_flapping() {
        // /t/-/ɾ/ acoustically ~0.5; overlay drops it to 0.10.
        assert!((sub_cost("t", "ɾ") - 0.10).abs() < EPS);
    }

    #[test]
    fn wce_overlay_fires_on_cot_caught() {
        assert!((sub_cost("ɑ", "ɔ") - 0.05).abs() < EPS);
    }

    #[test]
    fn similarity_is_one_for_identity() {
        assert_eq!(similarity("kæt", "kæt"), 1.0);
        assert_eq!(similarity("", ""), 1.0);
    }

    #[test]
    fn similarity_separates_madgab_pair_from_decoy() {
        let target = "ɪtsdʒʌstəstupɪdgeɪm";
        let clue   = "hɪtsdʒʌstɪsduphɪdkeɪm";
        let decoy  = "jɔrmʌðɝwɛrzsneɪkɝz";
        let s_clue  = similarity(target, clue);
        let s_decoy = similarity(target, decoy);
        assert!(s_clue - s_decoy >= 0.2,
                "clue {s_clue} not >= decoy {s_decoy} + 0.2");
    }

    #[test]
    fn boundary_indel_is_essentially_free() {
        // Same phonemes, repositioned word boundary. The pure-phoneme
        // contents are identical; confusion should reflect that.
        assert!(distance("ɪts dʒʌst", "ɪt sdʒʌst") < 0.05);
    }

    #[test]
    fn weak_phoneme_indel_is_cheap() {
        // Inserting /h/ at the head should cost roughly WEAK_INDEL_COST.
        assert!((distance("stupɪd", "hstupɪd") - WEAK_INDEL_COST).abs() < 0.01);
    }
}
